File: 2 - Structrual Design Patterns/6 Flyweight/FlyweightPatternApplied.py
class TreeType:
    """Flyweight class that stores shared (intrinsic) data for trees."""
    _instances = {}

    def __new__(cls, name, color, texture, height):
        key = (name, color, texture, height)
        if key not in cls._instances:
            cls._instances[key] = super().__new__(cls)
        return cls._instances[key]

    def __init__(self, name, color, texture, height):
        self.name = name
        self.color = color
        self.texture = texture
        self.height = height

class Tree:
    """Class that stores unique (extrinsic) data for trees."""
    def __init__(self, tree_type, location_x, location_y):
        self.tree_type = tree_type
        self.location_x = location_x
        self.location_y = location_y

    def display(self):
        print(f"Tree: {self.tree_type.name}, Color: {self.tree_type.color}, "
              f"Height: {self.tree_type.height}, Location: ({self.location_x}, {self.location_y})")

# Example usage of Flyweight pattern
forest = [
    Tree(TreeType("Oak", "Green", "Rough", 20), 5, 10),
    Tree(TreeType("Pine", "Dark Green", "Smooth", 30), 15, 30),
    Tree(TreeType("Oak", "Green", "Rough", 20), 50, 60),
    Tree(TreeType("Pine", "Dark Green", "Smooth", 30), 25, 80),
]

for tree in forest:
    tree.display()

File: 2 - Structrual Design Patterns/6 Flyweight/FlyweightPatternApplied.ts
class TreeType {
    private static _instances: { [key: string]: TreeType } = {};

    // Properties of the TreeType class
    public name: string;
    public color: string;
    public texture: string;
    public height: number;

    private constructor(name: string, color: string, texture: string, height: number) {
        this.name = name;
        this.color = color;
        this.texture = texture;
        this.height = height;
    }

    // Flyweight pattern: only one instance of a given type will be created
    public static getInstance(name: string, color: string, texture: string, height: number): TreeType {
        const key = `${name}-${color}-${texture}-${height}`;
        if (!this._instances[key]) {
            this._instances[key] = new TreeType(name, color, texture, height);
        }
        return this._instances[key];
    }
}

class Tree {
    private treeType: TreeType;
    private locationX: number;
    private locationY: number;

    constructor(treeType: TreeType, locationX: number, locationY: number) {
        this.treeType = treeType;
        this.locationX = locationX;
        this.locationY = locationY;
    }

    public display(): void {
        console.log(`Tree: ${this.treeType.name}, Color: ${this.treeType.color}, ` +
            `Height: ${this.treeType.height}, Location: (${this.locationX}, ${this.locationY})`);
    }
}

// Example usage of Flyweight pattern
const forest: Tree[] = [
    new Tree(TreeType.getInstance("Oak", "Green", "Rough", 20), 5, 10),
    new Tree(TreeType.getInstance("Pine", "Dark Green", "Smooth", 30), 15, 30),
    new Tree(TreeType.getInstance("Oak", "Green", "Rough", 20), 50, 60),
    new Tree(TreeType.getInstance("Pine", "Dark Green", "Smooth", 30), 25, 80),
];

// Displaying the trees in the forest
forest.forEach(tree => tree.display());
